### PR TITLE
lockstep refactoring of `KernelInsertParticles`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- rewrite kernel with the lockstep programming model
- add documentation

[HowTo lockstep programming](http://picongpu.readthedocs.io/en/dev/prgpatterns/lockstep.html)

## Tests

- [x] default KHI

### CODE REVIEW

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 1]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
     }
 };
 
-
+/** copy particles from exchange buffer into the border of the of the simulation
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: double `of the`

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 2]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
...
+ * @tparam T_numWorkers number of workers
+ */
+template< uint32_t T_numWorkers >
 struct KernelInsertParticles
 {
-    template<class ParBox, class BORDER, class Mapping>
-    DINLINE void operator()( ParBox pb,
-                                           ExchangePopDataBox<vint_t, BORDER, Mapping::Dim - 1 > border,
-                                           Mapping mapper ) const
+    /** copy particles from exchange buffer into the border of the of the simulation
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: double `of the`

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 3]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
...
+        PMACC_SMEM(
+            frame,
+            FramePtr
+        );
+        PMACC_SMEM(
+            elementCount,
+            int
+        );
+        PMACC_SMEM(
+            tmpBorder,
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: I think you called this `borderChunk` in #2086 now, same here?

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 4]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
...
-            tmpBorder = border.get(blockIdx.x,superCell);
-            elementCount = tmpBorder.getSize( );
-            if ( elementCount > 0 )
+        using MasterOnly = IdxConfig<
+            1,
+            numWorkers
+        >;
+
+        /* compressed index of the the supercell
+         * can be uncompressed with `DataSpaceOperations < > ::extend()`
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: weird spaces

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 5]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
...
-            parDestFull[multiMask_] = 1;
-            auto parSrc = ((tmpBorder[threadIdx.x])[0]);
-            /*we know that source has no multiMask*/
-            auto parDest = deselect<multiMask>(parDestFull);
-            assign( parDest, parSrc );
-        }
-        /*if this syncronize fix the kernel crash in spezial cases,
-         * I can't tell why.
+
+         /* loop over all particles in the frame */
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: //

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 6]
```diff
@@ -779,69 +779,156 @@ struct KernelBashParticles
...
+                    parDestFull[ multiMask_ ] = 1;
+                    auto parSrc = ( (tmpBorder[ linearIdx ] )[ 0 ] );
+                    /*we know that source has no multiMask*/
+                    auto parDest = deselect<multiMask>( parDestFull );
+                    assign( parDest, parSrc );
+                }
+            }
+        );
+
+        /** @bug This synchronize fix the kernel crash in special cases,
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: - fix*es*
- *a* kernel crash

**Comment by psychocoderHPC** [comment]
updated

**Pull request merged**